ang/prospect/services/view-pledge-contribution-case-action.service.js

```diff
@@ -0,0 +1,107 @@
+  var module = angular.module('prospect');
+
+  module.service('ViewPledgeContributionCaseAction', ViewPledgeContributionCaseAction);
+
+  function ViewPledgeContributionCaseAction ($location, crmApi, ProspectGlobalValues) {
+    var prospectConverted = {
+      isConverted: false,
+      paymentInfo: {
+        payment_completed: false,
+        pledge_balance: false,
+        payment_url: false,
+        payment_entity: false
+      }
+    };
+
+    /**
+     * Refresh Data for the Service
+     *
+     * @param {array} cases
+     */
+    this.refreshData = function (cases) {
+      if (cases[0]) {
+        var caseID = cases[0].id;
+
+        checkIfConvertedToProspect(caseID)
+          .then(function () {
+            getPaymentInfo(caseID);
+          });
+      }
+    };
+
+    /**
+     * Checks if the Action is allowed
+     *
+     * @param {Object} action
+     * @param {Array} cases
+     * @return {boolean}
+     */
+    this.isActionAllowed = function (action, cases) {
+      var isActionAllowed = action.type === 'pledge'
+        ? prospectConverted.paymentInfo.payment_entity === 'pledge'
+        : prospectConverted.paymentInfo.payment_entity === 'contribute';
+
+      return checkIfProspectingCaseTypeCategory() &&
+        prospectConverted.isConverted && isActionAllowed;
+    };
+
+    /**
+     * Click event handler for the Action
+     *
+     * @param {Array} cases
+     * @param {Object} action
+     * @param {Function} callbackFn
+     */
+    this.doAction = function (cases, action, callbackFn) {
+      var contactID = cases[0].client[0].contact_id;
+      var url = prospectConverted.paymentInfo.payment_url + '&cid=' + contactID;
+
+      return url;
+    };
+
+    /**
+     * Checks if the Case is Converted to Prospect
+     *
+     * @param {String/Int} caseID
+     * @return {Promise}
+     */
+    function checkIfConvertedToProspect (caseID) {
+      return crmApi('ProspectConverted', 'get', {
+        'sequential': 1,
+        'prospect_case_id': caseID
+      }).then(function (caseData) {
+        prospectConverted.isConverted = caseData.count > 0;
+      });
+    }
+
+    /**
+     * Gets the payment information
+     *
+     * @param {String/Int} caseID
+     */
+    function getPaymentInfo (caseID) {
+      if (prospectConverted.isConverted) {
+        crmApi('ProspectConverted', 'getpaymentinfo', {
+          'prospect_case_id': caseID
+        }).then(function (paymentInfo) {
+          prospectConverted.paymentInfo = paymentInfo;
+        });
+      }
+    }
+
+    /**
+     * Check if Case Type Category is Prospecting.
+     *
+     * @return {Boolean}
+     */
+    function checkIfProspectingCaseTypeCategory () {
+      var filtersQueryParams = $location.search().cf;
+      if (filtersQueryParams) {
+        var caseTypeCategory = JSON.parse(filtersQueryParams).case_type_category;
+
+        return caseTypeCategory === ProspectGlobalValues.caseTypeCategory;
+      }
+    }
+  }
+})(angular, CRM.$, CRM._);
```

Comment:
After checking all the code and seeing how `checkIfConvertedToProspect` and `checkIfProspectingCaseTypeCategory` I'd try to put those intro a service.

`checkIfProspectingCaseTypeCategory` should be on a service of its own, but I think it needs a bit of rework because it's not clear it does this check from the URL params.

I'd have a service like this instead:

```js
function getCaseTypeCategoryFromUrlParams() {
  var filtersQueryParams = $location.search().cf;

  if (!filtersQueryParams) {
    return;
  }

  return JSON.parse(filtersQueryParams).case_type_category
}
```

And use this service like this:

```js
getCaseTypeCategoryFromUrlParams() === ProspectGlobalValues.caseTypeCategory;
```

Reply:
for `checkIfConvertedToProspect` I'd refactor it like this:

```js
function ProspectConvertedService() {
  this.getProspectIsConverted = getProspectIsConverted;
  this. getPaymentInfo = getPaymentInfo;

  function getProspectIsConverted(caseId) {
     return crmApi('ProspectConverted', 'get', {
        sequential: 1,
        prospect_case_id: caseId
      }).then(function (caseData) {
        return caseData.count > 0;
      });
  }

  function getPaymentInfo (caseId) {
    crmApi('ProspectConverted', 'getpaymentinfo', {
      prospect_case_id: caseId
    }).then(function (paymentInfo) {
      return paymentInfo;
    });
  }
}
```

Then on the actual action:

```js
const isProspectConverted = false;
const prospectConvertedPaymentInfo = {};

// ...

this.refreshData = function (cases) {
  if (!cases[0]) {
    return;
  }

  ProspectConvertedService.getProspectIsConverted(cases[0].id)
    .then(function (isConverted) {
      isProspectConverted = isConverted;

      return isConverted
        ?  ProspectConvertedService.getPaymentInfo(cases[0].id)
        :  {};
    })
    .then(function (paymentInfo) {
      prospectConvertedPaymentInfo = paymentInfo;
    });
}
```

Reply:
Good tip. Fixed.